eureka2-server/src/main/java/com/netflix/eureka2/server/transport/tcp/discovery/TcpDiscoveryHandler.java

```diff
@@ -43,17 +52,74 @@
+
     private final EurekaCommonConfig config;
     private final SourcedEurekaRegistry<InstanceInfo> registry;
     private final EurekaServerMetricFactory metricFactory;
+    private final Subscription healthStatusSubscription;
+
+    private volatile boolean afterBootstrap;
 
     @Inject
     public TcpDiscoveryHandler(EurekaCommonConfig config,
                                SourcedEurekaRegistry registry,
+                               EurekaHealthStatusAggregator systemHealthStatus,
                                EurekaServerMetricFactory metricFactory) {
+        this(config, registry, systemHealthStatus, metricFactory, Schedulers.computation());
+    }
+
+    public TcpDiscoveryHandler(EurekaCommonConfig config,
+                               SourcedEurekaRegistry registry,
+                               EurekaHealthStatusAggregator systemHealthStatus,
+                               EurekaServerMetricFactory metricFactory,
+                               Scheduler scheduler) {
         this.config = config;
         this.registry = registry;
         this.metricFactory = metricFactory;
+        this.healthStatusSubscription = systemHealthStatus.healthStatus()
+                .filter(new Func1<HealthStatusUpdate<EurekaHealthStatusAggregator>, Boolean>() {
+                    @Override
+                    public Boolean call(HealthStatusUpdate<EurekaHealthStatusAggregator> update) {
+                        return update.getStatus() == Status.UP;
+                    }
+                })
+                .take(1)
+                .doOnError(new Action1<Throwable>() {
+                    @Override
+                    public void call(Throwable error) {
+                        logger.error("Healthcheck status stream terminated with an error. Interest connection " +
+                                "will not be accepted yet. Re-subscribing shortly", error);
+                    }
+                })
+                .retryWhen(new RetryStrategyFunc(RETRY_INTERVAL_MS, scheduler), scheduler)
+                .subscribe(
+                        new Subscriber<HealthStatusUpdate<EurekaHealthStatusAggregator>>() {
+                            @Override
+                            public void onCompleted() {
+                                if (!afterBootstrap) {
+                                    logger.error("Healthcheck status stream completed, while it is not yet marked as up.");
+                                }
+                            }
+
+                            @Override
+                            public void onError(Throwable e) {
+                                // This should never happen
+                                logger.error("Healthcheck stream terminated with an error and went out of retry loop", e);
+                            }
+
+                            @Override
+                            public void onNext(HealthStatusUpdate<EurekaHealthStatusAggregator> update) {
```

Comment:
I would keep logic simple for now. Also even if status is not UP anymore, it might be very risky to start discarding connections, unless we have a notion of total/partial failures.